### PR TITLE
Web3 provider types

### PR DESCRIPTION
**Pull request by joshstevens19**
`web3-providers` types written with full tests.

### CODE REVIEW

File: packages/web3-providers/types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    constructor(host: string, options: HttpProviderOptions);
+    send(payload: any, callback: () => void): void;
+    disconnect(): void;
+}
+
+export class IpcProvider {
+    constructor(path: string, net: net.Server);
+    addDefaultEvents(): void;
+    reconnect(): void;
+    send(payload: any, callback: () => void): void;
```
Comment (nivida) [inline packages/web3-providers/types/index.d.ts]: you could type the payload argument with the JSON-RPC v2 spec if you like.

File: packages/web3-providers/types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    constructor(providersPackageFactory: ProvidersModuleFactory);
+    resolve(provider: provider, net: net.Server): provider | Error;
+}
+
+export class JsonRpcResponseValidator {
+    static isValid(response: JsonRpcMappersPayload[] | JsonRpcMappersPayload): boolean;
+    static isResponseItemValid(response: JsonRpcMappersPayload): boolean;
+}
+
+export type provider = HttpProvider | IpcProvider | WebsocketProvider;
```
Comment (nivida) [inline packages/web3-providers/types/index.d.ts]: provider could also be of type string

File: packages/web3-providers/types/tests/http-provider-adapter-test.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+const options = {
+    timeout: 20000,
+    headers: [
+        {
+            name: 'Access-Control-Allow-Origin', value: '*'
+        }
+    ]
+};
+const httpProvider = new HttpProvider('http://localhost:8545', options);
+// $ExpectType HttpProviderAdapter
```
Comment (nivida) [inline packages/web3-providers/types/tests/http-provider-adapter-test.ts:35]: add a break above

**Review by nivida — CHANGES_REQUESTED** [review]

**Comment by joshstevens19** [comment]
@nivida perfect feedback thanks - all ready to be merged now

**Comment by joshstevens19** [comment]
@nivida perfect feedback thanks - can you check again and if happy merge it

**Pull request merged**